### ISSUE TITLE
The shape of vector attention map

### ISSUE THREAD

**Issue by Liu-Feng** [lucidrains/point-transformer-pytorch#2]
Firstly, thanks for your awesome work!!!
I have a confusion that if the vector attention can modulate individual feature channel, the attention map should have a axis whose dimension is same as the channel number of cooresponding value (V). An based on the Eq. (2) in the paper,  if the shape of query, key and value is [batch_size, num_point, num_dim],
the shape of vector attention map may should be [batch_size, num_points, num_points, num_dim]. Moreover, the "sim" which represent the vector attention map without softmax in your code, may should be calculate by "self.attn_mlp(qk_rel) + rel_pos_emb " based on the Eq.(2).

**Comment by lucidrains** [comment]
@Liu-Feng Hi Feng! Admittedly I wrote this repository and forgot to follow up and double check if it is correct

Do you mean to say it should be like https://github.com/lucidrains/point-transformer-pytorch/pull/3/files

**Comment by Liu-Feng** [comment]
@lucidrains Thanks!! If the last dimension of attn_mlp is dim other 1, I think it cloud be used to modulate individual feature channel. But I cannot confirm which axis that the softmax should conduct on. Thanks for your reply!

**Comment by lucidrains** [comment]
@Liu-Feng i'm pretty sure the softmax will be across the similarities of queries against all the keys (dimension `j`)

I've merged the PR, so do let me know if this work or doesn't work in your training

Thank you!

**Comment by Liu-Feng** [comment]
> against

Thanks for your reply!!! I have not train the model. I am trying to construct the PT via Tensorflow.

**Comment by lucidrains** [comment]
@Liu-Feng hello! how's your tensorflow port going? are you more certain that you had the correct hunch here?

**Comment by Liu-Feng** [comment]
Hollo, I carefully read the Point transformer, and the vector attention is conducted on a local path of the point cloud( which pointed out in the paper). 

If the number of downed-sampled points is N, and the dim of the learned feature is D, the B represents the batch size, thus, the feature shape is BND, and the shape of grouped feature is BNKD, where K is number of points in each local patch(K of KNN). 

Thus the attention may be conducted between BND and BNKD, the shape of Query , Key and Value are BND, BNKD and BNKD, respectively. The attention weight after substraction and mapping is BNKD(BND-BNKD, like the relative coordinates used in point cloud grouping in Pointnet++). Then the hadamard product is made bettwen attention weight(BNKD) and value(BNKD), following by the reduce_sum made in K dim(axis=2). Therefore, the shape of output feature is BND(with squeeze in dim 2). 

In this way, the wector weight could refine every channel in Value. But it works like the attention, other than the self-attention. All these steps are based on my understandings of the paper. 
The real workflow of the Point Transformer may difference from my understandings, and the truth cloud be uncover after the author open the source code.

By the way, some interesting things may be found the in orignal vector attention paper( the code of which is opened), which is written by the sample author of Point transformer.    

Have a good day!!
